sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/InitReplayingStreamObserver.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.sgv2.common.grpc;
+
+import io.grpc.stub.StreamObserver;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A stream observer that starts its life accumulating incoming elements; it only starts streaming
+ * them (via {@link #onNextReady(Object)} once {@link #markReady()} is called.
+ */
+public abstract class InitReplayingStreamObserver<T> implements StreamObserver<T> {
+
+  // This is based on the "queue-drain" pattern
+  // https://github.com/ReactiveX/RxJava/wiki/Implementing-custom-operators-(draft)#the-queue-drain-approach
+  private final Queue<T> elements = new ConcurrentLinkedQueue<>();
+  private final AtomicReference<ObserverState> state = new AtomicReference<>(ObserverState.INITIAL);
```

Comment:
This is complicated but I can't think of a simpler solution.

I suspect calls to `onNext()` are always happening from the same Netty event loop, but `markReady()` definitely isn't, so we need proper synchronization.